x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/lucene/LuceneMinMaxOperator.java

```diff
@@ -0,0 +1,175 @@
+    public Page getCheckedOutput() throws IOException {
+        if (isFinished()) {
+            assert remainingDocs <= 0 : remainingDocs;
+            return null;
+        }
+        final long start = System.nanoTime();
+        try {
+            final LuceneScorer scorer = getCurrentOrLoadNextScorer();
+            // no scorer means no more docs
+            if (scorer == null) {
+                remainingDocs = 0;
+            } else {
+                final LeafReader reader = scorer.leafReaderContext().reader();
+                final Query query = scorer.weight().getQuery();
+                if (query == null || query instanceof MatchAllDocsQuery) {
+                    final PointValues pointValues = reader.getPointValues(fieldName);
+                    // only apply shortcut if we are visiting all documents, otherwise we need to trigger the search
+                    // on doc values as that's the order they are visited without push down.
+                    if (pointValues != null && pointValues.getDocCount() >= remainingDocs) {
+                        final Bits liveDocs = reader.getLiveDocs();
+                        if (liveDocs == null) {
+                            // In data partitioning, we should have already collected it
+                            if (scorer.position() == 0) {
+                                seen = true;
+                                result = numberType.evaluate(result, numberType.fromPointValues(pointValues));
+                                if (remainingDocs != NO_LIMIT) {
+                                    remainingDocs -= pointValues.getDocCount();
+                                }
+                            }
+                            scorer.markAsDone();
+                        }
+                    }
+                }
+                if (scorer.isDone() == false) {
+                    // could not apply shortcut, trigger the search
+                    final NumericDocValues values = numberType.multiValueMode(reader.getSortedNumericDocValues(fieldName));
+                    final LeafCollector leafCollector = new LeafCollector() {
+                        @Override
+                        public void setScorer(Scorable scorer) {}
+
+                        @Override
+                        public void collect(int doc) throws IOException {
+                            assert remainingDocs > 0;
+                            remainingDocs--;
+                            if (values.advanceExact(doc)) {
+                                seen = true;
+                                result = numberType.evaluate(result, values.longValue());
+                            }
+                        }
+                    };
+                    scorer.scoreNextRange(leafCollector, reader.getLiveDocs(), remainingDocs);
+                }
+            }
+
+            Page page = null;
+            // emit only one page
+            if (remainingDocs <= 0 && pagesEmitted == 0) {
```

Comment:
I want to confirm - this exists so that we can be called over and over and over again and we will transfer control back to the main loop after each slice.

Reply:
I must say I just follow the implementation in LuceneCountOperator: https://github.com/elastic/elasticsearch/blob/21ccde548f07479637d2831cbc2058356582efda/x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/lucene/LuceneCountOperator.java#L140

I don't really follow all this protection, we have `doneCollecting`, `remainingDocs` and `pagesEmitted` which sort of signal if we are done or not?

Reply:
I believe it's about yielding after performing "some work" and letting the driver have another iteration so it can update status and check cancellation and things. I don't know precisely where it came from, but that's my guess of it.